src/App.tsx

```diff
@@ -15,35 +16,107 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  alphabetically = 'alphabetically',
+  byLength = 'byLength',
+  default = '',
+}
+
+type SortObj = {
+  sortField: SortType;
+  isReversed?: boolean;
+};
+
+function getPreparedGoods(
+  goods: string[],
+  { sortField, isReversed = false }: SortObj,
+) {
+  const preparedGoods = [...goods];
+
+  preparedGoods.sort((good1, good2) => {
+    switch (sortField) {
+      case SortType.alphabetically:
+        return good1.localeCompare(good2);
+
+      case SortType.byLength:
+        return good1.length - good2.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState<SortType>(SortType.default);
+  const [isReversed, setIsReversed] = useState(false);
+  const visibledGoods = getPreparedGoods(goodsFromServer, {
+    sortField,
+    isReversed,
+  });
+
+  const reset = () => {
+    setIsReversed(false);
+    setSortField(SortType.default);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-info', {
+            'is-light': sortField !== SortType.alphabetically,
+          })}
+          onClick={() => setSortField(SortType.alphabetically)}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-success', {
+            'is-light': sortField !== SortType.byLength,
+          })}
+          onClick={() => setSortField(SortType.byLength)}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-warning', {
+            'is-light': !isReversed,
+          })}
+          onClick={() => setIsReversed(!isReversed)}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortField || isReversed) && (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={reset}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
         <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
+          {visibledGoods.map(good => (
+            <li data-cy="Good" key={good}>
+              {good}
+            </li>
+          ))}
         </ul>
```

Comment:
There is a nested `ul` element inside another `ul` without an intermediary `li` element, which is not valid HTML. You should remove the unnecessary `ul` element to correct the structure. Also, for the list items being rendered, ensure that the `key` prop is unique and stable, such as using an ID if available. This is important for rendering lists in React as per the [CODE KNOWLEDGE] checklist.